File: Cpp/1008_Construct_Binary_Search_Tree_from_Preorder_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        TreeNode* root = nullptr;
        stack<TreeNode*> parents;
        
        for(auto& num : preorder){
            TreeNode* node = new TreeNode(num);
            if(!root) root = node;
            
            if(parents.empty()){
                parents.push(node);
            } else {
                if(parents.top()->val > node->val) {
                    parents.top()->left = node;
                    parents.push(node);
                } else {
                    TreeNode* parent = nullptr;
                    while(!parents.empty() && parents.top()->val < node->val) {
                        parent = parents.top();
                        parents.pop();
                    }
                    
                    parent->right = node;
                    parents.push(node);
                }
            }
        }
        
        return root;
    }
};

File: Kotlin/1008_Construct_Binary_Search_Tree_from_Preorder_Traversal.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun bstFromPreorder(preorder: IntArray): TreeNode? {
        val root = TreeNode(preorder[0])
        for(i in 1 until preorder.size) {
            setNode(root, preorder[i])
        }
        
        return root
    }
    
    private fun setNode(node: TreeNode, value: Int) {
        if(value > node.`val`) {
            if(node.right == null) node.right = TreeNode(value)
            else setNode(node.right, value)
        }else {
            if(node.left == null) node.left = TreeNode(value)
            else setNode(node.left, value)
        }
    }
}